repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AssetPreProcessor.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.atlas.authorize.AtlasPrivilege;
+import org.apache.atlas.exception.AtlasBaseException;
+import org.apache.atlas.model.instance.*;
+import org.apache.atlas.repository.graphdb.AtlasVertex;
+import org.apache.atlas.repository.store.graph.v2.EntityGraphRetriever;
+import org.apache.atlas.repository.store.graph.v2.EntityMutationContext;
+import org.apache.atlas.type.AtlasTypeRegistry;
+import org.apache.atlas.utils.AtlasPerfMetrics;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+import static org.apache.atlas.repository.Constants.*;
+
+public class AssetPreProcessor implements PreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(AssetPreProcessor.class);
+
+    private EntityMutationContext context;
+    private AtlasTypeRegistry typeRegistry;
+    private EntityGraphRetriever entityRetriever;
+
+    public AssetPreProcessor(AtlasTypeRegistry typeRegistry, EntityGraphRetriever entityRetriever) {
+        this.typeRegistry = typeRegistry;
+        this.entityRetriever = entityRetriever;
+    }
+
+    @Override
+    public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context,
+                                  EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("AssetPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+        this.context = context;
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        switch (operation) {
+            case CREATE:
+                processCreateAsset(entity, vertex);
+                break;
+            case UPDATE:
+                processUpdateAsset(entity, vertex);
+                break;
+        }
+    }
+
+    private void processCreateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+          validateDomainAssetLinks(entity);
+          AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+          isAuthorized(sourceEntity);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+
+    private void processUpdateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+            validateDomainAssetLinks(entity);
+            AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+            isAuthorized(sourceEntity);
+        }
```

Comment:
can you create another method with this section like processDomainAttribute() & call this processDomainAttribute() from both processCreateAsset() as well as processUpdateAsset()